script.py

```diff
@@ -0,0 +1,186 @@
+import argparse
+
+spark = SparkSession.builder.getOrCreate()
+
+program_description = '''
+This script retrieves the structures in complex with the molecules from parquet.
+'''
+
+parser = argparse.ArgumentParser(add_help=True, description=program_description)
+
+parser.add_argument('-p', '--parquet', help='Path to parquet folder', default=None, metavar='parquet_file_path', type=str, required=True)
+parser.add_argument('-u', '--unichem', help='Path to unichem src1src3.txt file', default=None, metavar='unichem_file_path', type=str, required=True)
+parser.add_argument('-e', '--ensembl', help='Path to pdb_chain_ensembl.csv file', default=None, metavar='ensembl_file_path', type=str, required=True)
+
+args = parser.parse_args()
+
+
+def get_structure(pdb_id):
+    """
+    Scrap the list of structures in complex with each molecule from PDBe API
+    Function to apply to a df column
+    --------------
+    Keyword arguments
+        pdb_id: id of the molecule for whom we want to know the structures associated
+    --------------
+    Return
+        data[pdb_id] : data is a dictionary, we want the value which is the list of structures
+    """
+    url = f'https://www.ebi.ac.uk/pdbe/api/pdb/compound/in_pdb/{pdb_id}'
+    response = requests.get(url)
+    try:
+        data = response.json()
+        return data[pdb_id]
+    except:
+        if len(response.json()) == 0:
+            return []
+
+
+def create_pdb_target_gene_df(path_id_file, unichem_molecule_struct_spark_df):
+
+    pdb_chain_ensembl = spark.read.csv(path_id_file, sep=',', header=True, comment='#')
+    pdb_chain_ensembl = pdb_chain_ensembl.select('PDB', 'GENE_ID').distinct()
+
+    # print('----- STRUCTURE & TARGET -----')
+    # pdb_chain_ensembl.show()
+
+    exploded_df = (unichem_molecule_struct_spark_df
+                   .select('MOLECULE_PDB_ID',
+                           'MOLECULE_CHEMBL_ID',
+                           f.explode(unichem_molecule_struct_spark_df.STRUCTURE_ID)))
+
+    # print('----- STRUCTURE & TARGET EXPLODED -----')
+    # exploded_df.show()
+
+    chain_ensembl_struct_mol_joined = (pdb_chain_ensembl
+                                       .join(exploded_df, pdb_chain_ensembl["PDB"] == exploded_df["col"])
+                                       .drop('col'))
+
+    return chain_ensembl_struct_mol_joined
+
+
+def main():
+
+    # TIMER 1 START
+    start_1 = timeit.default_timer()
+
+    # ----- GET MOLECULE ID -----
+    # Molecule DataFrame
+    molecule_df = (
+        spark.read
+        .parquet(args.parquet)
+        .select(f.col('id'))
+        .withColumnRenamed('id', 'MOLECULE_CHEMBL_ID')
+    )
+
+    # ----- GET PDB ID -----
+    # Unichem molecules DataFrame
+    unichem_df = spark.read.csv(args.unichem, sep=r'\t', header=True)
+
+    # Join Unichem and Molecule DataFrame
+    unichem_molecule_df = (molecule_df
+                           .join(unichem_df, unichem_df["From src:'1'"] == molecule_df['MOLECULE_CHEMBL_ID'])
```

Comment:
You don't need to drop the column you joined the two tables with if you are already renamed them. And in general it is easier to deal with the columns if you rename the columns when you read the files.

Reply:
#### This is what I have if I dont select/drop a column:

```
+-----------------+------------------+------------------+---------------+
|             name|CHEMBL_MOLECULE_ID|CHEMBL_MOLECULE_ID|PDB_COMPOUND_ID|
+-----------------+------------------+------------------+---------------+
|       MIGALASTAT|      CHEMBL110458|      CHEMBL110458|            DGJ|
|   (S)-Fluoxetine|     CHEMBL1169388|     CHEMBL1169388|            SFX|
|        RUCAPARIB|     CHEMBL1173055|     CHEMBL1173055|            RPB|
|    CHEMBL1231606|     CHEMBL1231606|     CHEMBL1231606|            C4F|
|    CHEMBL1232381|     CHEMBL1232381|     CHEMBL1232381|            DTB|
|      PF-04691502|     CHEMBL1234354|     CHEMBL1234354|            ML9|
```

#### I replaced the drop by:

```python
# Join Unichem and Molecule DataFrame
unichem_molecule_df = (molecule_df.alias("a")
                       .join(unichem_df.alias("b"), unichem_df["CHEMBL_MOLECULE_ID"] == molecule_df['CHEMBL_MOLECULE_ID'])
                       ).select("name", "a.CHEMBL_MOLECULE_ID", "PDB_COMPOUND_ID")
```

Is it better?

Reply:
If you join two tables by columns with the same name you can simplify this expression:
```python
unichem_df["CHEMBL_MOLECULE_ID"] == molecule_df['CHEMBL_MOLECULE_ID']
```
to this:
```python
on='CHEMBL_MOLECULE_ID'
```
In the above case columns will not be duplicated. Also if you need to join by multiple columns you can say:
```python
on=['CHEMBL_MOLECULE_ID', 'targetId']
```

Nicer, isn't it?